### ISSUE TITLE
Incorrect Spin Hall Conductivity Calculation using WannierBerri and VASP

### ISSUE THREAD

**Issue by ma455173220** [wannier-berri/wannier-berri#274]
Hi All,

I am currently working on utilizing Wannierberri to calculate the spin Hall conductivity of various materials. As part of my preliminary efforts, I conducted benchmarking by performing calculations on Pt using VASP. Although the band structure closely aligns with experimental data near the Fermi level, as convincingly demonstrated in the attached document "Band structure - wannier90_VASP_W90_cmp.pdf " under VASP folder, I have encountered an anomaly. The computed spin Hall conductivity appears to be much smaller and inaccurate (data in WannierBerri-SHC.xlsx under VASP folder).
 
To further investigate this issue, I calculated the spin Hall conductivity using the "wannier90_hr.dat" file and Wanniertools. The calculated results are consistent with the reported data (data in WannierTools-SHC.xlsx under VASP folder). A direct comparison between the results obtained from WannierTools and WannierBerri can be found in "Compare WannierTools and WannierBerri.png" under VASP folder:
![Compare WannierTools and WannierBerri](https://github.com/wannier-berri/wannier-berri/assets/42956329/5b8da070-8d7b-4382-ae44-b9f08c2eed6c)

I also performed calculations using a significantly denser k-point mesh (15x15x15). The data obtained from WannierBerri appears to be more accurate compared to the results obtained with a smaller k-point mesh. However, it's worth noting that the magnitude of the results at the Fermi level is still considerably lower:
![Compare WannierTools and WannierBerri - kmesh-15x15x15](https://github.com/wannier-berri/wannier-berri/assets/42956329/7aa0aa53-cc87-4d3a-81e9-1b320cbf00e2)

Additionally, I followed the example provided by Wannierberri, which utilizes QE, and in this case, the spin Hall conductivity was correctly computed for both WannierTools and WannierBerri (data can be found under QE folder). Therefore, I suspect that the discrepancy in the spin Hall conductivity value, specifically when using Wannierberri in conjunction with VASP, may be attributed to the "wannier90.spn" file.
 
If it is not too much trouble, would you be willing to review my input settings to identify any potential issues? All the input and output files are attached. (Notes: In WannierTools, the unit of the Spin Hall Conductivity (SHC) is (ħ/e) S/cm, which means you should divide WannierBerri's results by 100 to obtain the comparable values. Additionally, please note that the data from WannierTools has already been adjusted so that the Fermi level is shifted to 0.)

Thank you immensely for your valuable time and assistance.
 
Best Regards,
Hongyang

[WAVECAR.zip](https://github.com/wannier-berri/wannier-berri/files/12554576/WAVECAR.zip)
[wannier90_mmn.zip](https://github.com/wannier-berri/wannier-berri/files/12554577/wannier90_mmn.zip)
[Pt-VASP-SHC-WannierBerri.zip](https://github.com/wannier-berri/wannier-berri/files/12554579/Pt-VASP-SHC-WannierBerri.zip)

**Comment by stepan-tsirkin** [comment]
@jaemolihm @manxkim @jhryoo  , any ideas?

**Comment by jaemolihm** [comment]
If I understood correctly, the bug arises only with VASP, not QE. Could you check whether your calculations are affected by VASP's Wannier interface?
https://www.vasp.at/wiki/index.php/Known_issues

**Comment by ma455173220** [comment]
Yes, QE-WannierBerri works fine. I utilized VASP version 6.3.1, so according to the website, the VASP-Wannier interface should work smoothly. Furthermore, as demonstrated in the attached figure, employing WannierTools and the wannier90_hr.dat file generated from the identical VASP-Wannier process produces accurate outcomes.

**Comment by stepan-tsirkin** [comment]
note that [vaspspn](https://docs.wannier-berri.org/en/master/docs/vaspspn.html)
> computes 𝑆𝑚𝑛𝐪 based on the normalized pseudo-wavefunction read from the WAVECAR file. Note that the use of pseudo-wavefunction instead of the full PAW ([Blöchl 1994](https://journals.aps.org/prb/abstract/10.1103/PhysRevB.50.17953)) wavefunction is an approximation, which however in practice gives a rather accurate interpolation of spin.

This "good in practice" is probably not so much tested. What it was good for is to interpolate spin in a material whithout Kramers degeneracy (breaking either inversion or time-reversal), e.g. to plot figures like  [Fig11 here](https://journals.aps.org/prb/abstract/10.1103/PhysRevB.97.035158)
![image](https://github.com/wannier-berri/wannier-berri/assets/22172977/999e877f-c186-4b3f-a6c6-fd3a58a21740)

But I actually never tested it for spin-hall, especially in Kramers-degenerate systems. I am actually not sure if it is a correct thing to do. 

One thing to check - take a material without Kramers degeneracy (say, Fe or Te) and the spin along lines like in the [tutorial ](https://tutorial.wannier-berri.org/tutorials/1_basic/tutorial-wb-basic-solution.html#Interpolation-on-a-path) from .spn file obtained with vaspspn, and compare it with what you get from PROCAR, or with QE calculation. If the figures do make sense. Also for iron, the total [spin](https://docs.wannier-berri.org/en/master/docs/calculators.html#wannierberri.calculators.static.Spin) may be computed  and compared to VASP self-consistent calculation.  

If all that makes sense, then probably vaspspn is not a proper thing to use for spin-hall calculations.

I must say, VASP has a very poor interface to Wannier, and this is the reason that I personally avoid using it anymore. `.spn` file is a trivial and a must-have thing, but we are waiting for years to have it.

**Comment by ma455173220** [comment]
Thank you, Dr. Stepan, for your response! I'll definitely test out the approach you suggested. Actually, I'm in the process of learning Quantum Espresso (QE) precisely because of the challenging interface between VASP and Wannier90. Just one more question: Regarding your point about WannierTools making an approximation that each spin Wannier function is an eigenvalue of the spin_z operator, I was wondering if it would be feasible to incorporate this into WannierBerri as well?

**Comment by stepan-tsirkin** [comment]
> I was wondering if it would be feasible to incorporate this into WannierBerri as well?

Yes, it must be quite straightforward. In fact there is such thing for the spin matrix: 

https://github.com/wannier-berri/wannier-berri/blob/070dc9b1b88716f71d7a1b0d0593b6c71908628a/wannierberri/system/system.py#L285-L304

In addition one should set the matrices `SR`, `SH` and `SHR` under this approximation, which are normally set from .spn file here :

https://github.com/wannier-berri/wannier-berri/blob/070dc9b1b88716f71d7a1b0d0593b6c71908628a/wannierberri/system/system_w90.py#L144-L151

**Comment by stepan-tsirkin** [comment]
Oups, I just realized that the `set-spin` function is wrong. I will correct it soon.

**Comment by ma455173220** [comment]
It sounds a little complex to me... but I'll definitely give it a shot. Thanks so much for your assistance!

**Comment by LiFu1997** [comment]
Dear all,
Similar situation also happened to me when I use vaspspn utility. When I use VASP+Wannierberri to calculate the SHC, the SHC of Pt is about 900. and I use the VASP + postw90.x(using vaspspn utilityt o generate wannier90.spn file for wannier90 from WAVECAR file generated by VASP), but the result is also obout 900. I agree with ma455173220, may be is the wannier90.spn  problem. When Dr. Stepan corrected it. Please let me know.Thanks very much.

Best regards
Fu

**Comment by ma455173220** [comment]
Dear all,

I'd like to share my recent experience in case it may help others facing a similar issue in the future. I successfully resolved the problem by recompiling VASP with the patch available at https://github.com/Chengcheng-Xiao/VASP2WAN90_v2_fix. After recompiling, it's important to note that you should run VASP using only a SINGLE CPU core to generate mmn, amn, and spn. I encountered the same underestimation issue when attempting to generate mmn and amn using multiple CPUs. While I'm not an expert, I suspect this might be related to the use of openmpi or some other factor. Therefore, if you're struggling to obtain correct results, I recommend trying this patch and SINGLE CPU approach. Good luck!

Regards,
Hongyang

**Comment by stepan-tsirkin** [comment]
@ma455173220 , Thank you for sharing your experience.

With what version of VASP did you use the patch? Because it writes that 

>THIS FIX ONLY WORKS WITH VASP v5.4.4.pl2

But above you write that you use  VASP version 6.3.1

**Comment by ma455173220** [comment]
Dear Dr. Stepan,

I recompiled VASP with v5.4.4.pl2. I didn't try 6.3.1 for this patch.

Regards,
Hongyang

**Comment by stepan-tsirkin** [comment]
@ma455173220 , did you also try vaspspn with  VASP  v5.4.4.pl2 ?

Just, vaspspn was developed with that version in mind, and I am not sure if anything changed in writing the WAVECAR file in later versions. 

Just to make sure?

**Comment by ma455173220** [comment]
Dear Dr. Stepan,

I just tried with VASP v5.4.4.pl2 but it gave an error like this:
![screenshot](https://github.com/wannier-berri/wannier-berri/assets/42956329/2fc10caa-a18d-475a-9821-34494d2a493b)

Regards,
Hongyang

**Comment by ma455173220** [comment]
Dear Dr. Stepan,

I conducted a test by using the .spn file generated from 'vaspspn,' while keeping all other files, including .mmn, .amn, .chk, and .eig, exactly the same. The results still indicate underestimation, similar to the previous calculation, approximately around 1.5E+03, when compared to the results obtained using the .spn file generated from the patch.

Regards,
Hongyang

**Comment by stepan-tsirkin** [comment]
Ok, so the patch is a more proper thing then vaspspn.py. Better use it.

**Comment by sharmashivamm16** [comment]
> Hi All,
> 
> I am currently working on utilizing Wannierberri to calculate the spin Hall conductivity of various materials. As part of my preliminary efforts, I conducted benchmarking by performing calculations on Pt using VASP. Although the band structure closely aligns with experimental data near the Fermi level, as convincingly demonstrated in the attached document "Band structure - wannier90_VASP_W90_cmp.pdf " under VASP folder, I have encountered an anomaly. The computed spin Hall conductivity appears to be much smaller and inaccurate (data in WannierBerri-SHC.xlsx under VASP folder).
> 
> To further investigate this issue, I calculated the spin Hall conductivity using the "wannier90_hr.dat" file and Wanniertools. The calculated results are consistent with the reported data (data in WannierTools-SHC.xlsx under VASP folder). A direct comparison between the results obtained from WannierTools and WannierBerri can be found in "Compare WannierTools and WannierBerri.png" under VASP folder: ![Compare WannierTools and WannierBerri](https://user-images.githubusercontent.com/42956329/266476926-5b8da070-8d7b-4382-ae44-b9f08c2eed6c.png)
> 
> I also performed calculations using a significantly denser k-point mesh (15x15x15). The data obtained from WannierBerri appears to be more accurate compared to the results obtained with a smaller k-point mesh. However, it's worth noting that the magnitude of the results at the Fermi level is still considerably lower: ![Compare WannierTools and WannierBerri - kmesh-15x15x15](https://user-images.githubusercontent.com/42956329/266476373-7aa0aa53-cc87-4d3a-81e9-1b320cbf00e2.png)
> 
> Additionally, I followed the example provided by Wannierberri, which utilizes QE, and in this case, the spin Hall conductivity was correctly computed for both WannierTools and WannierBerri (data can be found under QE folder): ![Compare WannierTools and WannierBerri](https://user-images.githubusercontent.com/42956329/266477227-d6e3844d-6a88-41d0-89af-2ffc429c6840.png)
> 
> Therefore, I suspect that the discrepancy in the spin Hall conductivity value, specifically when using Wannierberri in conjunction with VASP, may be attributed to the "wannier90.spn" file.
> 
> If it is not too much trouble, would you be willing to review my input settings to identify any potential issues? All the input and output files are attached. (Notes: In WannierTools, the unit of the Spin Hall Conductivity (SHC) is (ħ/e) S/cm, which means you should divide WannierBerri's results by 100 to obtain the comparable values. Additionally, please note that the data from WannierTools has already been adjusted so that the Fermi level is shifted to 0.)
> 
> Thank you immensely for your valuable time and assistance.
> 
> Best Regards, Hongyang
> 
> [WAVECAR.zip](https://github.com/wannier-berri/wannier-berri/files/12554576/WAVECAR.zip) [wannier90_mmn.zip](https://github.com/wannier-berri/wannier-berri/files/12554577/wannier90_mmn.zip) [Pt-VASP-SHC-WannierBerri.zip](https://github.com/wannier-berri/wannier-berri/files/12554579/Pt-VASP-SHC-WannierBerri.zip)

The unit of SHC that we get from Wanniertools and wannier90 is (ħ/e) S/cm, what are the units of SHC in wannierberri ?

**Comment by sharmashivamm16** [comment]
The unit of SHC that we get from Wanniertools and wannier90 is (ħ/e) S/cm, what are the units of SHC in wannierberri ?

**Comment by ma455173220** [comment]
> The unit of SHC that we get from Wanniertools and wannier90 is (ħ/e) S/cm, what are the units of SHC in wannierberri ?

In Wannierberri, the unit should be (ħ/e) S/m, which is 100 times larger than Wanniertools.

Regards,
Hongyang
